server/src/metrics/metric_batching.rs

```diff
@@ -0,0 +1,328 @@
+    where
+        T: std::fmt::Display,
+    {
+        ClientMetricsEnv {
+            app_name: format!("app_name_{}", toggle_id),
+            feature_name: format!("toggle-{}", toggle_id),
+            environment: "development".into(),
+            timestamp: DateTime::parse_from_rfc3339("1867-11-07T12:00:00Z")
+                .unwrap()
+                .with_timezone(&Utc),
+            yes: 1,
+            no: 1,
+            variants: HashMap::new(),
+            metadata: MetricsMetadata {
+                platform_name: None,
+                platform_version: None,
+                sdk_version: None,
+                sdk_type: None,
+                yggdrasil_version: None,
+            },
+        }
+    }
+
+    fn make_impact_metric_env<T>(toggle_id: T) -> ImpactMetricEnv
+    where
+        T: std::fmt::Display,
+    {
+        ImpactMetricEnv {
+            app_name: format!("app_name_{}", toggle_id),
+            environment: "development".into(),
+            impact_metric: ImpactMetric {
+                name: format!("impact-{}", toggle_id),
+                help: "Impact metric for testing".into(),
+                r#type: "counter".into(),
+                samples: vec![],
+            },
+        }
+    }
+
+    #[test_case(10, 100, 1; "10 apps 100 toggles. Will not be split")]
+    #[test_case(1, 10000, 26; "1 app 10k toggles, will be split into 26 batches")]
+    #[test_case(1000, 1000, 7; "1000 apps 1000 toggles, will be split into 7 batches")]
+    #[test_case(500, 5000, 15; "500 apps 5000 toggles, will be split into 15 batches")]
+    #[test_case(5000, 1, 20; "5000 apps 1 metric will be split")]
+    fn splits_successfully_into_sendable_chunks(apps: u64, toggles: u64, batch_count: usize) {
+        let apps: Vec<ClientApplication> =
+            (1..=apps).map(|app_id| make_client_app(app_id)).collect();
+
+        let toggles: Vec<ClientMetricsEnv> = (1..=toggles)
+            .map(|toggle_id| make_metrics_env(toggle_id))
+            .collect();
+
+        let cache = MetricsCache::default();
+        for app in apps.clone() {
+            cache.applications.insert(
+                ApplicationKey {
+                    app_name: app.app_name.clone(),
+                    instance_id: app.instance_id.clone().unwrap_or_else(|| "unknown".into()),
+                },
+                app,
+            );
+        }
+        cache.sink_metrics(&toggles);
+        let batches = cache.get_appropriately_sized_batches();
+
+        assert_eq!(batches.len(), batch_count);
+        assert!(batches.iter().all(sendable));
+        // Check that we have no duplicates
+        let applications_sent_count = batches.iter().flat_map(|b| b.applications.clone()).count();
+
+        assert_eq!(applications_sent_count, apps.len());
+
+        let metrics_sent_count = batches.iter().flat_map(|b| b.metrics.clone()).count();
+        assert_eq!(metrics_sent_count, toggles.len());
+    }
+
+    fn execute_test(
```

Comment:
Factored out so that prop_test regressions are copy pastable into single tests for study (see last test in the file for an example)